### ISSUE TITLE
INSERT INTO - no results?

### ISSUE THREAD

**Issue by Eugenius1991** [hgourvest/node-firebird#135]
Hello

I am trying to write a function that lets me insert a value into the firebird database. The query works well, only I get no callback to tell me that the insert went well. Has anybody encountered similar problems?

`        pool.get(function(err, db) {
            if (err) {
                console.log(err)
                res.send(403)
            }
            else {
                var date = moment(req.body.date, "DD/MM/YYYY")
                var values = " VALUES ('" + date.format("MM/DD/YYYY") + "','Requested','" + req.body.type + "','" + req.body.description + "','" + req.body.memo +"')"
                console.log(values)
                var query = 'INSERT INTO "Features" (FEATURE_REQUESTDATE, FEATURE_STATUS, FEATURE_TYPE, FEATURE_DESCRIPTION, FEATURE_MEMO)' +
                    values + ' RETURNING FEATURE_ID'
                
                db.query( query , function(err, result) {
                    if (result) {
                        console.log(result) //no results to log?
                        res.status(200)
                        res.send('ok')
                    }
                    if (err) {
                        console.log(err)
                        res.status(403)
                        res.send('error')
                    }
                })
                    db.detach();
                }

            })`

**Comment by ericlovescode** [comment]
Try using db.execute.

**Comment by Eugenius1991** [comment]
Same behaviour with db.execute.

**Comment by paulish** [comment]
This is a bug in node-firebird communication with firebird-server.

Instead of call of execute and fetch from it node-firebird must call execute2 without fetch.

The same issue has been fixed in pdo-firebird: https://github.com/php/php-src/commit/13ffa88e1f62c4407c55e1b82c299fc452211662#diff-7fb3606b83568fc8c83cbd98b64890c1

Hope this info is helpful.

**Comment by ericlovescode** [comment]
How critical is this bug? I'm going live soon with a node-firebird-based
system.

**Comment by paulish** [comment]
This is critical only if you have Firebird 3 (works fine with 2.5). At the same time you can workaround issue in most of cases by 2 queries: 1) select id from generator / procedure 2) insert with previously selected and saved value.

**Comment by semiautomatix** [comment]
Can it be that this is still an issue?

**Comment by felipezacani** [comment]
> Can it be that this is still an issue?

Yes, it is.

**Comment by albanirneves** [comment]
Any updates about this?

**Comment by mariuz** [comment]
The callback does fire for INSERTs (with no result rows); if you saw it never called on that era's driver, that was the response-queue bug class fixed in 2.x. For confirmation of what happened, v2.12.0's `withMeta: true` returns `affectedRows`/`recordCounts` for every DML — [docs](https://github.com/hgourvest/node-firebird#result-metadata-and-affected-rows-withmeta). And `INSERT ... RETURNING ID` gives you the new row's values directly. Closing as answered.
